plutus-ir/src/Language/PlutusIR/Parser.hs

```diff
@@ -0,0 +1,344 @@
+import           Control.Monad.State          hiding (fail)
+
+import qualified Language.PlutusCore          as PLC
+import qualified Language.PlutusCore.Constant as PLC
+import           Language.PlutusIR            as PIR
+import           Text.Megaparsec              hiding (State, parse)
+
+import           Data.ByteString.Internal     (c2w)
+import qualified Data.ByteString.Lazy         as BS
+import           Data.Char
+import           Data.Foldable
+import qualified Data.Map                     as M
+import qualified Data.Text                    as T
+import           Data.Word
+import           GHC.Natural
+
+import           Text.Megaparsec.Char
+import qualified Text.Megaparsec.Char.Lexer   as Lex
+
+data ParserState = ParserState { identifiers :: M.Map String PLC.Unique, counter :: PLC.Unique }
+    deriving (Show)
+
+data PIRParseError = Overflow Natural Integer
+                   | UnexpectedKeyword String
+                   | InternalError String
+                   deriving (Eq, Ord, Show)
+
+instance ShowErrorComponent PIRParseError where
+    showErrorComponent (Overflow sz i) = "Integer overflow: " ++ show i ++ " does not fit in " ++ show sz ++ " bytes"
+    showErrorComponent (UnexpectedKeyword kw) = "Keyword " ++ kw ++ " used as identifier"
+    showErrorComponent (InternalError cause) = "Internal error: " ++ cause
+
+initial :: ParserState
+initial = ParserState M.empty (PLC.Unique 0)
+
+intern :: MonadState ParserState m => String -> m PLC.Unique
+intern n = do
+    st <- get
+    case M.lookup n (identifiers st) of
+        Just u -> return u
+        Nothing -> do
+            let counter' = PLC.Unique $ (PLC.unUnique $ counter st) + 1
+                identifiers' = M.insert n counter' $ identifiers st
+            put $ ParserState identifiers' counter'
+            return counter'
+
+type Parser = ParsecT PIRParseError String (State ParserState)
+
+parse :: String -> Parser a -> String -> Either (ParseError Char PIRParseError) a
+parse filename p s = evalState (runParserT p filename s) initial
+
+-- No block comments, no line comments
+whitespace :: Parser ()
+whitespace = Lex.space space1 (Lex.skipLineComment "--") (Lex.skipBlockCommentNested "{-" "-}")
+
+-- Tokens
+-- TODO: move this to separate module?
+
+lexeme :: Parser a -> Parser a
+lexeme = Lex.lexeme whitespace
+
+symbol :: String -> Parser String
+symbol = Lex.symbol whitespace
+
+lparen :: Parser String
+lparen = symbol "("
+rparen :: Parser String
+rparen = symbol ")"
+
+lbracket :: Parser String
+lbracket = symbol "["
+rbracket :: Parser String
+rbracket = symbol "]"
+
+lbrace :: Parser String
+lbrace = symbol "{"
+rbrace :: Parser String
+rbrace = symbol "}"
+
+inParens :: Parser a -> Parser a
+inParens = between lparen rparen
+
+inBrackets :: Parser a -> Parser a
+inBrackets = between lbracket rbracket
+
+inBraces :: Parser a -> Parser a
+inBraces = between lbrace rbrace
+
+builtins :: [(String, PLC.BuiltinName)]
+builtins =
```

Comment:
We have `Language.PlutusCore.Lexer.Type.allBuiltinNames`, I believe that's what you want to use here, because the set of builtins changes over time and you'll get automatic updates of the parser if you just handle all builtins generically.